[processor/logdedup] Panic in getLogKey function

### Component(s)

_No response_

### What happened?

## Description
A user reported this issue in Slack, due to a lack of clear documentation on the `conditions` being combined with a logical `or`. They ended up attempting to refer to an include_field expected to be nested in body on log where body was not a map.

The counter/getLogKey function has a switch where it retrieves a pcommon.Map from either the body or attributes, but it provides no default behavior (throw error if your include_field does not start with body or attributes), and does not propagate an error if attempting to address body when body is not a map.

## Panic
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xb19630]

goroutine 225 [running]:
go.opentelemetry.io/collector/pdata/pcommon.Map.Get({0x0?, 0x0?}, {0xc000926b75, 0xb})
        go.opentelemetry.io/collector/pdata@v1.32.0/pcommon/map.go:66 +0x30
github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor.getKeyValue({0x0?, 0x0?}, {0xc000f42270, 0x1, 0x1})
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor@v0.126.0/counter.go:235 +0x65
github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor.getLogKey({0xc0013b0800?, 0xc000b696f0?}, {0xc001427500?, 0x47b10f?, 0x91011c22ff64a711?})
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor@v0.126.0/counter.go:211 +0x3d4
github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor.(*scopeAggregator).Add(0xc000b98180, {0xc0013b0800?, 0xc000b696f0?})
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor@v0.126.0/counter.go:144 +0x45
github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor.(*resourceAggregator).Add(0xc000b98120, {0xc0009ca1c0?, 0xc000b696f0?}, {0xc0013b0800?, 0xc000b696f0?})
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor@v0.126.0/counter.go:123 +0x17b
github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor.(*logAggregator).Add(0xc000fa7640, {0xc00177e438?, 0xc000b696f0?}, {0xc0009ca1c0?, 0xc000b696f0?}, {0xc0013b0800?, 0xc000b696f0?})
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor@v0.126.0/counter.go:91 +0x1bc
github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor.(*logDedupProcessor).aggregateLog(0xc001653020, {0xc0013b0800?, 0xc000b696f0?}, {0xc0009ca1c0?, 0xc000b696f0?}, {0xc00177e438?, 0xc000b696f0?})
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor@v0.126.0/processor.go:131 +0xed
github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor.(*logDedupProcessor).ConsumeLogs.func1({0xc0013b0800?, 0xc000b696f0?})
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor@v0.126.0/processor.go:111 +0x2e5
go.opentelemetry.io/collector/pdata/plog.LogRecordSlice.RemoveIf({0xc0009ca200?, 0xc000b696f0?}, 0xc00143dc38)
        go.opentelemetry.io/collector/pdata@v1.32.0/plog/generated_logrecordslice.go:131 +0x72
github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor.(*logDedupProcessor).ConsumeLogs(0xc001653020, {0xcdace20, 0xc00166f200}, {0xc001673440?, 0xc000b696f0?})
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdedupprocessor@v0.126.0/processor.go:98 +0x356
go.opentelemetry.io/collector/processor/processorhelper.NewLogs.func1({0xcdace20, 0xc00166f200}, {0xc001673440?, 0xc000b696f0?})
        go.opentelemetry.io/collector/processor/processorhelper@v0.126.0/logs.go:66 +0x28b
go.opentelemetry.io/collector/consumer.ConsumeLogsFunc.ConsumeLogs(...)
        go.opentelemetry.io/collector/consumer@v1.32.0/logs.go:27
go.opentelemetry.io/collector/processor/batchprocessor.(*batchLogs).export(0xc000ed46c0?, {0xcdace20?, 0xc00166f200?}, {0xc001673440?, 0xc000b696f0?})
        go.opentelemetry.io/collector/processor/batchprocessor@v0.126.0/batch_processor.go:543 +0x2f
go.opentelemetry.io/collector/processor/batchprocessor.(*shard[...]).sendItems(0xc001855f34, 0x0?)
        go.opentelemetry.io/collector/processor/batchprocessor@v0.126.0/batch_processor.go:262 +0x8b
go.opentelemetry.io/collector/processor/batchprocessor.(*shard[...]).startLoop(0xce1a180)
        go.opentelemetry.io/collector/processor/batchprocessor@v0.126.0/batch_processor.go:222 +0x1c5
created by go.opentelemetry.io/collector/processor/batchprocessor.(*shard[...]).start in goroutine 1
        go.opentelemetry.io/collector/processor/batchprocessor@v0.126.0/batch_processor.go:186 +0x8a
```

### Collector version

0.126.0

### Environment information

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")


### OpenTelemetry Collector configuration

```yaml

```

### Log output

```shell

```

### Additional context

_No response_

--- a/.chloggen/jeffalder_fix-dedup.yaml
+++ b/.chloggen/jeffalder_fix-dedup.yaml
@@ -1,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: logdedupprocessor
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Taught logdedupprocessor not to panic if the dedup field source was not found or not a map
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [40204]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: [user]

--- a/processor/logdedupprocessor/counter.go
+++ b/processor/logdedupprocessor/counter.go
@@ -192,25 +192,17 @@ func getScopeKey(scope pcommon.InstrumentationScope) uint64 {
 
 // getLogKey creates a unique hash for the log record to use as a map key.
 // If dedupFields is non-empty, it is used to determine the fields whose values are hashed.
+// If no dedupFields are found in the log record, all fields are hashed.
 func getLogKey(logRecord plog.LogRecord, dedupFields []string) uint64 {
 	if len(dedupFields) > 0 {
 		var opts []pdatautil.HashOption
 
 		for _, field := range dedupFields {
 			parts := splitField(field)
-			var m pcommon.Map
-			switch parts[0] {
-			case bodyField:
-				if logRecord.Body().Type() == pcommon.ValueTypeMap {
-					m = logRecord.Body().Map()
+			if m, ok := getMap(logRecord, parts[0]); ok {
+				if value, ok := getKeyValue(m, parts[1:]); ok {
+					opts = append(opts, pdatautil.WithString(value.AsString()))
 				}
-			case attributeField:
-				m = logRecord.Attributes()
-			}
-
-			value, ok := getKeyValue(m, parts[1:])
-			if ok {
-				opts = append(opts, pdatautil.WithString(value.AsString()))
 			}
 		}
 
@@ -225,6 +217,20 @@ func getLogKey(logRecord plog.LogRecord, dedupFields []string) uint64 {
 		pdatautil.WithString(logRecord.SeverityNumber().String()),
 		pdatautil.WithString(logRecord.SeverityText()),
 	)
+}
+
+func getMap(logRecord plog.LogRecord, leadingPart string) (pcommon.Map, bool) {
+	switch leadingPart {
+	case bodyField:
+		if logRecord.Body().Type() == pcommon.ValueTypeMap {
+			return logRecord.Body().Map(), true
+		}
+	case attributeField:
+		return logRecord.Attributes(), true
+	}
+
+	var m pcommon.Map
+	return m, false
 }
 
 func getKeyValue(valueMap pcommon.Map, keyParts []string) (pcommon.Value, bool) {

--- a/processor/logdedupprocessor/counter_test.go
+++ b/processor/logdedupprocessor/counter_test.go
@@ -294,6 +294,38 @@ func Test_getLogKey(t *testing.T) {
 				require.Equal(t, expectedMulti, getLogKey(logRecord, []string{"body.dedup_key", "attributes.dedup_key"}))
 			},
 		},
+		{
+			desc: "getLogKey hashes full message if dedup key is body-based and no body was provided",
+			testFunc: func(t *testing.T) {
+				logRecord := plog.NewLogRecord()
+				logRecord.Attributes().PutStr("str", "attr str")
+
+				expected := pdatautil.Hash64(
+					pdatautil.WithMap(logRecord.Attributes()),
+					pdatautil.WithValue(logRecord.Body()),
+					pdatautil.WithString(logRecord.SeverityNumber().String()),
+					pdatautil.WithString(logRecord.SeverityText()),
+				)
+
+				require.Equal(t, expected, getLogKey(logRecord, []string{"body.dedup_key"}))
+			},
+		},
+		{
+			desc: "getLogKey hashes full message if dedup key is body-based and body is not map type",
+			testFunc: func(t *testing.T) {
+				logRecord := plog.NewLogRecord()
+				logRecord.Body().SetStr("hello, this is a message body string")
+
+				expected := pdatautil.Hash64(
+					pdatautil.WithMap(logRecord.Attributes()),
+					pdatautil.WithValue(logRecord.Body()),
+					pdatautil.WithString(logRecord.SeverityNumber().String()),
+					pdatautil.WithString(logRecord.SeverityText()),
+				)
+
+				require.Equal(t, expected, getLogKey(logRecord, []string{"body.dedup_key"}))
+			},
+		},
 	}
 
 	for _, tc := range testCases {

